### ISSUE TITLE
Rest parameter tuples don't propagate generics

### ISSUE THREAD

**Issue by JakeTunaley** [microsoft/TypeScript#25256]
**TypeScript Version:**  3.0.0-dev.20180626

**Search Terms:** tuple rest spread variadic call apply generic

**Code**

```ts
function generic<T, U> (a: T, b: U): [T, U] {
	return [a, b];
}

function apply<A extends any[], B> (fn: (...args: A) => B, args: A): B {
	return fn.apply(undefined, args);
}

let x = apply(generic, [7, 'my string']);
x[0] = 'another string';
```
**Bug:**
The apply call infers A = `[any, any]` and B = `[any, any]`, instead of A and B = `[number, string]`. This gives `x` a type of `[any, any]`, and thus the assignment to `x[0]` does not generate an error.

**Related Issues:**
PR [#24897](https://github.com/Microsoft/TypeScript/pull/24897)

**Comment by mhegazy** [comment]
This is a design limitation in the current system. We currently do not have a way to traffic free type variables and make them manifest in a different signature.

**Comment by mhegazy** [comment]
//cc @DanielRosenwasser

**Comment by RyanCavanaugh** [comment]
This is now `string | number`, which is a big step up from `any`. I don't think we can improve much beyond that without rethinking how parameter lists get handled in a pretty fundamental way.

**Comment by droganov** [comment]
Struggling to type a validation rule factory because of this limitation

The factory

```js
export type ValidateRule<I, O> = (input: I, context: ValidationContext) => O
export type SyncRule<I, O> = (context: ValidationContext) => (input: I) => O

export const createRule =
  <I, O>(validate: ValidateRule<I, O>): SyncRule<I, O> =>
  context =>
  input =>
    validate(input, context)
```

Calling
```js
let result = myRule(context)(1234)
```

Infers as `unkown`
